LSP: Move workspace/configuration from nvim-lspconfig to core

mjlbach opened the pull request:
Half-addresses https://github.com/neovim/neovim/issues/13615

cc @teto @ckipp01

mjlbach committed:
LSP: Add helper function from accessing language server settings from nvim-lspconfig

mjlbach committed:
LSP: Move workspace/configuration handler from nvim-lspconfig to core

mfussenegger commented on runtime/lua/vim/lsp/handlers.lua:
```diff
@@ -149,6 +149,28 @@ M['workspace/applyEdit'] = function(_, _, workspace_edit)
+--@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_configuration
+M['workspace/configuration'] = function(err, _, params, client_id)
+  local client = vim.lsp.get_client_by_id(client_id)
+  if err then error(tostring(err)) end
```
```suggestion
  if err then error(vim.inspect(err)) end
```

Afaik `error` is a table with the code and message, so `tostring(err))` would return `table: 0x...`.

mfussenegger reviewed (commented):
Nice to see this getting added to neovim core.


I guess the documentation of the parameters for `vim.lsp.start_client()` should also be updated to mention the new `settings` property.

mfussenegger commented on runtime/lua/vim/lsp/handlers.lua:
```diff
@@ -149,6 +149,28 @@ M['workspace/applyEdit'] = function(_, _, workspace_edit)
 
+--@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_configuration
+M['workspace/configuration'] = function(err, _, params, client_id)
+  local client = vim.lsp.get_client_by_id(client_id)
```
Should there also be a `nil` check for the `client` similar to how it's done here?

https://github.com/neovim/neovim/commit/fb90b6ec9803fdbd603b1d5f1685583218c18909#diff-728d3ae352b52f16b51a57055a3b20efc4e992efacbf1c34426dfccbba30037cR73

mjlbach commented:
@mfussenegger I think I addressed everything in https://github.com/neovim/neovim/pull/13649/commits/938841950a1dc80eaa89195faffa0c095c7e8ed0

I will squash commits before merge

ckipp01 commented:
Thanks for addressing this @mjlbach!

I see that this only actually addresses the `workspace/configuration` handler, but not the `workspace/didChangeConfiguration`. You editing the docs to say that it will be sent once, but I don't see that anywhere? I believe that this should be tied into the `on_init` and right after the initialize, shoot the `workspace/didChangeconfiguration` to the server. imo that part is actually more critical than the `workspace/configuration`. It should be able to be something really simple like this: https://github.com/scalameta/nvim-metals/blob/3aaf999a4ae8c91afd33d3db90a31459459044db/lua/metals/setup.lua#L168-L170

mjlbach commented:
@ckipp01 I'm slightly confused. Reading back over your issue, I agree we should move the handlers to core. However, in core you *can* currently use the following to update settings.

```lua
client.notify('workspace/didChangeConfiguration', {
            settings = settings;
          })
```
All we do here is wrap this with additional checks for valid settings. Are you saying this helper function should be moved into core as well?

https://github.com/neovim/nvim-lspconfig/blob/4f431b992482d9edbe4ccf65170e6304503614a9/lua/lspconfig/configs.lua#L148-L156

ckipp01 commented:
> @ckipp0 Reading back over your issue, I agree we should move the handlers to core. However, in core you _can_ currently use the following to update settings.
> 
> ```lua
> client.notify('workspace/didChangeConfiguration', {
>             settings = settings;
>           })
> ```
> 
> All we do with `client.workspace_did_change_configuration` [here](https://github.com/neovim/nvim-lspconfig/blob/4f431b992482d9edbe4ccf65170e6304503614a9/lua/lspconfig/configs.lua#L148-L156) is wrap this with additional checks for valid settings.
> 
> Just to clarify, are you saying this helper function should be moved into core as well, and exposed as a method of client as it is in nvim_lspconfig? I can change the documentation to refer to the client.notify approach otherwise.

Correct, it's for sure possible to do already, but it's not done automatically right away after initialize (in core). For people using this and not using nvim-lspconfg, I think this will be confusing. Since most lsp clients will send a `workspace/didChangeConfiguration` right after the initialize happens. So if we say that `settings` is part of config, and document it, but then don't actually send it to the server after initialize, then it could/will cause confusion, and also miss the main point of having this here. I'd argue that doing this automatically right after the initialize is the most important part since if it's not done, we force the user to do another manual step in order for this to work, when we can just take care of it. Does that make sense?

mjlbach commented:
> It's for sure possible to do already, but it's not done automatically right away after initialize (in core). 

My main concern is that I'm wondering if there is a situation where ovewriting the on_init call of `vim.lsp.start_client({config})` with a wrapped call that invokes `workspace/didChangeConfiguration` as we do in nvim-lspconfig might be undesirable in some cases, such as debugging servers. 

If that isn't the case, I can do as you ask and bring some of the wrapping logic from on_init in nvim-lspconfig to core. Another option might be to expose a helper function such as vim.lsp.util.wrap_on_init() that returns a wrapped on_init that can be passed  as part of config `vim.lsp.start_client({config})`

> So if we say that settings is part of config, and document it, but then don't actually send it to the server after initialize, then it could/will cause confusion, and also miss the main point of having this here. 

The main point of having this here is to respond to workspace/configuration handler.

> I'd argue that doing this automatically right after the initialize is the most important part since if it's not done, we force the user to do another manual step in order for this to work, when we can just take care of it.

Isn't this what nvim-lspconfig is for? 

I'm not opposed to doing what you're asking, I'm working on it right now to see what it would look like.

mfussenegger commented on runtime/doc/lsp.txt:
```diff
@@ -755,6 +755,12 @@ start_client({config})                                *vim.lsp.start_client()*
                                          array.
                     {handlers}         Map of language server method names to
                                        |lsp-handler|
+                    {settings}         Map with case-sensitive keys to
+                                       'workspace/configuration' event
+                                       responses. These are sent to the server
+                                       via `workspace/didChangeConfiguration`.
+                                       If settings are changed, the client must 
+                                       be notified with `client.workspace_did_change_configuration({settings})`.
```
```suggestion
                    {settings}         Map with language server specific
                                       settings. These are returned to the language server if
                                       requested via `workspace/configuration`. Keys are
                                       case-sensitive.
```


I'd trim this down to the `workspace/configuration` part for now given that the `didChangeConfiguration` part is still in nvim-lspconfig and not in core.

mfussenegger reviewed (commented):
I'd keep this PR focused on the `workspace/configuration` part and the changes look good to me so far. Maybe a test case would be great? Not sure if necessary.



Regarding the `didConfigurationChange` aspect. I'm not sure if it is correct to fire that event immediately within `on_init`? Based on the specification I'd think that the server is supposed to use `workspace/configuration` to request the configuration, and the client is supposed to notify the server if the configuration changes. That would be the case if neovim either had an option to change the settings during runtime, or if neovim used a configuration file for the settings and the user edited it.

Do servers really rely on an initial `didConfigurationChange` event, even if the configuration did not change?

mjlbach commented:
That was my intuition as well, I'll make the changes to the documentation and I'm (finally) figuring out how to write LSP test-cases.  I have a separate PR incoming that will expose a wrapper which allows you to wrap an on_init call with the didChangeConfiguration request.

mjlbach commented:
I'll try disabling the onChangeConfiguration even in on_init and see what breaks in a branch of nvim-lspconfig.

ckipp01 commented:
> My main concern is if there is a situation where ovewriting the on_init call of vim.lsp.start_client({config}) with a wrapped call that invokes workspace/didChangeConfiguration as we do in nvim-lspconfig might be undesirable/confusing in certain cases, such as debugging servers.

Understandable. I wouldn't fully overwrite it though, probably find a way to combine it with any `on_init` that the user could have given. The flip side to your concern would be having settings included in the config, then the user passing settings in, and then never having them applied. Imo this would cause even more confusion. Why isn't my settings being applied?

> Another option might be to expose a helper function such as vim.lsp.util.wrap_on_init() that returns a wrapped on_init that can be passed as part of config

Totally, this would be ideal I think.

> The main point of having this here is to respond to workspace/configuration handler.

Understood. This is probably an issue with how I phrased the initial issue. I view this and `workspace/didChangeConfiguration` going hand in hand since once you have the settings, these are the two that would use it.

> Based on the specification I'd think that the server is supposed to use workspace/configuration to request the configuration, and the client is supposed to notify the server if the configuration changes.

@mfussenegger I think in theory, that's how it can work, yes. However just picking 3 examples, coc.nvim, vim-lsp, and even nvim-lspconfig _all_ follow the pattern of sending a `workspace/didChangeConfiguration` right after the initialize. I believe that's a pretty standard way to have settings set. I know on the server side at least for us in Metal, we never actually use `workspace/configuration` afaik. We simply rely on the `didChangeConfiguration` being sent, and that has seemed to be the case for every client we work with so far.

> > I'd argue that doing this automatically right after the initialize is the most important part since if it's not done, we force the user to do another manual step in order for this to work, when we can just take care of it.

> Isn't this what nvim-lspconfig is for?

> I'm not opposed to doing what you're asking, I'm working on it right now to see what it would look like.

I think `nvim-lspconfig` is a separate concern. I think from the core point of view, the core parts of the lsp spec should be handled there, and then any plugin can utilize them, such as `nvim-lspconfig`. I'm totally in the minority here, but for example `nvim-metals` doesn't use `nvim-lspconfig`. I don't want to speak for him, but @mfussenegger is in the same situation with `nvim-jdtls`. So for people that aren't using `nvim-lspconfg` for their server specific extensions (which there may be more of in the future), they have to re-impliment the logic that is included in `nvim-lspconfig` if it's not included in core. So from a selfish point of view, I'd want to make sure that things that are central to the LSP configuration can at least be done with minimal work in core for users in my situation. Again, I'm totally the minority here since I'm going against the grain and providing a plugin that isn't reliant on `nvim-lspconfig` 😈.

mjlbach committed:
LSP: add test for workspace/configuration handler

mjlbach commented:
@mfussenegger The test is not correct, I'm having trouble figuring out how to trigger the lsp handler in the on_callback function (see commented out line 284).

I'm also a bit confused as to if we should be using request/expect_request here

mjlbach commented:
> this [adding the helper function to wrap on_init] would be ideal I think.

Awesome! I'll have the PR up in a couple days.

> Understood. This is probably an issue with how I phrased the initial issue. I view this and workspace/didChangeConfiguration going hand in hand since once you have the settings, these are the two that would use it.

No worries, I marked this as not resolving your issue so it will stay up. 

> I know on the server side at least for us in Metals, we never actually use workspace/configuration afaik. We simply rely on the didChangeConfiguration being sent, and that has seemed to be the case for every client we work with so far.

That's good to know, I think we should continuing sending this by default. 

> I'd argue that doing this automatically right after the initialize is the most important part since if it's not done, we force the user to do another manual step in order for this to work, when we can just take care of it.

Based on your earlier comments, I think we're in agreement that the wrapper + documentation would address this.

> I think nvim-lspconfig is a separate concern. I think from the core point of view, the core parts of the lsp spec should be handled there, and then any plugin can utilize them, such as nvim-lspconfig. I'm totally in the minority here, but for example nvim-metals doesn't use nvim-lspconfig. I don't want to speak for him, but @mfussenegger is in the same situation with nvim-jdtls. So for people that aren't using nvim-lspconfg for their server specific extensions (which there may be more of in the future), they have to re-impliment the logic that is included in nvim-lspconfig if it's not included in core. So from a selfish point of view, I'd want to make sure that things that are central to the LSP configuration can at least be done with minimal work in core for users in my situation. 

This is a valid point, I'll do what I can to implement things that will make life easier. I don't think nvim-lspconfig should be a hard dependency for writing a plugin that uses the neovim language client.

mfussenegger commented:
> However just picking 3 examples, coc.nvim, vim-lsp, and even nvim-lspconfig all follow the pattern of sending a workspace/didChangeConfiguration right after the initialize

Likely all copying from each other 😆


>  I know on the server side at least for us in Metals, we never actually use workspace/configuration afaik. We simply rely on the didChangeConfiguration being sent, and that has seemed to be the case for every client we work with so far.

Looking through

- https://github.com/microsoft/language-server-protocol/issues/972 and
- https://github.com/microsoft/language-server-protocol/issues/567

Specifically:

- https://github.com/microsoft/language-server-protocol/issues/567#issuecomment-448538082

I think relying on `didChangeConfiguration` to provide the full configuration is not the right thing to do for a server.

Either way, sounds like sending a `workspace/didChangeConfiguration` with all the settings on/after initialization is something that a client is allowed to do, and that it works best for most servers, so maybe that's the best option to go with.

ckipp01 commented:
> > However just picking 3 examples, coc.nvim, vim-lsp, and even nvim-lspconfig all follow the pattern of sending a workspace/didChangeConfiguration right after the initialize
> 
> Likely all copying from each other 😆

Exactly 😆 

> > I know on the server side at least for us in Metals, we never actually use workspace/configuration afaik. We simply rely on the didChangeConfiguration being sent, and that has seemed to be the case for every client we work with so far.
> 
> Looking through
> 
>     * [microsoft/language-server-protocol#972](https://github.com/microsoft/language-server-protocol/issues/972) and
> 
>     * [microsoft/language-server-protocol#567](https://github.com/microsoft/language-server-protocol/issues/567)
> 
> 
> Specifically:
> 
>     * [microsoft/language-server-protocol#567 (comment)](https://github.com/microsoft/language-server-protocol/issues/567#issuecomment-448538082)
> 
> 
> I think relying on `didChangeConfiguration` to provide the full configuration is not the right thing to do for a server.

Thanks for that last link, I actually haven't seen that.

> Either way, sounds like sending a `workspace/didChangeConfiguration` with all the settings on/after initialization is something that a client is allowed to do, and that it works best for most servers, so maybe that's the best option to go with.

Like so many things in LSP, this seems to be common behavior for servers/clients, so agreed. Probably best to stick to convention until a critical mass does change, and then adjust accordingly.

teto commented:
nvim-lspconfig should definitely be optional, the split between core and the configs was hard to do so now it's time to cleanup. 
It looks cleaner to set the correct options from the get go (especially if you deal with a remote LSP server) but this can be kept in mind and done after the refactoring, let's stick to what's known to work.

mjlbach commented:
Last issue with tests is for some reason settings is not being stored when set in on_init, meaning the handler fails.

mjlbach commented:
Ok, fixed the test! I think it's good now? I squashed and cleaned up the commit messages (one for the util, one for the moved handler, one for the test)

@mfussenegger and @teto Can you give the test a once over?

mjlbach commented on test/functional/plugin/lsp_spec.lua:
```diff
@@ -262,6 +262,48 @@ describe('LSP', function()
+            local result = exec_lua([=[
+              local method, params = ...
+              return require'vim.lsp.handlers'['workspace/configuration'](err, method, params, TEST_RPC_CLIENT_ID)]=], method, params)
+            client.notify('workspace/configuration', result)
```
Note, this is kinda an ugly hack that we can't directly return the handler's result to the fake-lsp

@mfussenegger I'm not sure fixing this is within the scope of this PR, but this would be a very useful utility function to have (unless I'm missing it)

mfussenegger replied:
Looks okay as it is to me, but I'm also not too familiar with the `test_rpc_server` stuff - maybe somebody else has inputs on whether there is a simpler way to do this.

mfussenegger commented on runtime/lua/vim/lsp/handlers.lua:
```diff
@@ -149,6 +149,32 @@ M['workspace/applyEdit'] = function(_, _, workspace_edit)
+--@see https://microsoft.github.io/language-server-protocol/specifications/specification-current/#workspace_configuration
+M['workspace/configuration'] = function(err, _, params, client_id)
+  local client = vim.lsp.get_client_by_id(client_id)
+  local client_name = client and client.name or string.format("id=%d", client_id)
+  if not client then
+    err_message("LSP[", client_name, "] client has shut down after sending the message")
+  end
```
```suggestion
  if not client then
    err_message("LSP[id=", client_id, "] client has shut down after sending the message")
  end
```

`client_name` was only used in the `not client` case, so it always evaluates to `id=<client_id>`.

mjlbach replied:
We were discussing this a bit in https://github.com/neovim/neovim/pull/13652

I'm good with standardizing on reporting the error by client_id, especially because there can be multiple clients in the log sharing the same client_name

mfussenegger replied:
My suggestion still keeps the same error reporting format as there was before - it just removed the `client and client.name` case which is dead code. `client.name` is never used. If the `client` is available then the `if` branch is skipped.

mjlbach replied:
Re-reading it now I see that we couldn't even get client_name if we wanted to for that error message. Thanks for the catch!

mfussenegger reviewed (commented):
Left a minor suggestion, overall looks good to me.

teto commented on runtime/lua/vim/lsp/handlers.lua:
```diff
@@ -149,6 +149,31 @@ M['workspace/applyEdit'] = function(_, _, workspace_edit)
+      local value = util.lookup_section(client.config.settings, item.section) or vim.NIL
+      -- For empty sections with no explicit '' key, return settings as is
+      if value == vim.NIL and item.section == '' then
+        value = client.config.settings or vim.NIL
```
shouldn't it return straightaway in that case ? since it covers all of the config.  It doesn't hurt to go on so it may be ok.

mjlbach replied:
I guess there is the corner case to worry about, where in the array of ConfigurationParams, the first param requests an item.section that is empty (adding client.config.settings to the table), and the second ConfigurationParams requests one that is not empty (adding a specific setting to the table).

something like:
```lua
params = {
    items = { 
        { section = ''}
        { section = 'python.analysis.autoSearchPaths'}
    }
}
```

I don't know if there *is* a server that does this (it would probably be a bug), but I think we should respond to the entire list of items.

teto commented:
is it ready to be merged ?

mjlbach commented:
Think so, if you agree with my comment

C3pa commented:
Would you mind updating the wiki: https://github.com/neovim/nvim-lspconfig/wiki/Project-local-settings ?

clason commented:
The Wiki is maintained by the nvim-lspconfig project, please open a new issue there if you think it needs changes.
